packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -235,15 +236,67 @@ class TextPainter {
   double get preferredLineHeight {
     if (_layoutTemplate == null) {
       final ui.ParagraphBuilder builder = ui.ParagraphBuilder(
-        _createParagraphStyle(TextDirection.rtl),
-      ); // direction doesn't matter, text is just a space
+        _createParagraphStyle(TextDirection.ltr),
+      ); // direction doesn't matter, text is just a space, ltr is more common.
       if (text?.style != null)
         builder.pushStyle(text.style.getTextStyle(textScaleFactor: textScaleFactor));
       builder.addText(' ');
       _layoutTemplate = builder.build()
         ..layout(ui.ParagraphConstraints(width: double.infinity));
     }
-    return _layoutTemplate.height;
+    return max(_layoutTemplate.height, _maxLineHeight);
+  }
+
+  // Tracks the max line height to prevent jitter in text box size.
+  //
+  // When a taller-than-existing glyph is typed, prevent the height from
+  // shrinking smaller than it.
+  double _maxLineHeight = 0;
+
+  /// Attempt to get the actual height of the line after layout.
+  ///
+  /// This returns the max height of the line that the character at [offset]
+  /// belongs in. When the paragraph has not been laid out, this returns the
+  /// estimate given by [preferredLineHeight] as a fallback.
+  ///
+  /// The [offset] parameter is the byte index of the text the height should
+  /// be queried at.
+  ///
+  /// This method may produce different heights than [preferredLineHeight] as
+  /// this is the actual measured height of the laid out text. This means that
+  /// the height of each line can vary depending on the maximum heights of
+  /// the characters in the line.
+  double preferredLineHeightAtOffset(int offset) {
+    if (_needsLayout) {
+      return preferredLineHeight;
+    }
+    final double height = _getHeightFromUpstream(offset)
+        ?? _getHeightFromDownstream(offset)
+        ?? preferredLineHeight;
+    _maxLineHeight = max(height, _maxLineHeight);
```

Comment:
it seems weird that calling `preferredLineHeightAtOffset` would have a side-effect. TextPainter should really be idempotent and path-independent; changing the string then reading a bunch of preferred heights then changing the string then reading one preferred heights should do nothing different than setting the string to the final value then reading that final preferred height.

Reply:
I'll see if I can make it so that the state is separate from the existing getter.